wgengine/magicsock/magicsock.go

```diff
@@ -2704,55 +2746,126 @@ func (et *endpointTracker) update(now time.Time, eps []tailcfg.Endpoint) (epsPlu
+	// possible, with space freed by the entries removed after expiry.
 	for _, ep := range eps {
 		et.addLocked(now, ep, until)
 	}
 
-	// Remove everything that has now expired.
-	et.removeExpiredLocked(now)
+	// Finally, add entries to the return array that aren't already there.
+	epsPlusCached = eps
+	for _, heap := range et.endpoints {
+		for _, ep := range heap {
+			// If the endpoint was in the input list, or has expired, skip it.
+			if inputEps.Contains(ep.endpoint.Addr) {
+				continue
+			} else if now.After(ep.until) {
+				// Defense-in-depth; should never happen since
+				// we removed expired entries above, but ignore
+				// it anyway.
+				continue
+			}
+
+			// We haven't seen this endpoint; add to the return array
+			epsPlusCached = append(epsPlusCached, ep.endpoint)
+		}
+	}
+
+	// Remove the oldest items from the heap
 	return epsPlusCached
 }
 
-// add will store the provided endpoint(s) in the cache for a fixed period of
-// time, and remove any entries in the cache that have expired.
+// extendLocked will update the expiry time of the provided endpoint in the
+// cache, if it is present. If it is not present, nothing will be done.
+//
+// et.mu must be held.
+func (et *endpointTracker) extendLocked(ep tailcfg.Endpoint, until time.Time) {
+	key := ep.Addr.Addr()
+	epHeap, found := et.endpoints[key]
+	if !found {
+		return
+	}
+
+	// Find the entry for this exact address; this loop is quick since we
+	// bound the number of items in the heap.
+	//
+	// TODO(andrew): this means we iterate over the entire heap once per
+	// endpoint; even if the heap is small, if we have a lot of input
+	// endpoints this can be expensive?
+	for i, entry := range epHeap {
+		if entry.endpoint == ep {
+			entry.until = until
+			heap.Fix(&epHeap, i)
+			mak.Set(&et.endpoints, key, epHeap)
+			return
+		}
+	}
+}
+
+// addLocked will store the provided endpoint(s) in the cache for a fixed
+// period of time, ensuring that the size of the endpoint cache remains below
+// the maximum.
 //
 // et.mu must be held.
 func (et *endpointTracker) addLocked(now time.Time, ep tailcfg.Endpoint, until time.Time) {
-	// If we already have an entry for this endpoint, update the timeout on
-	// it; otherwise, add it.
-	entry, found := et.cache[ep.Addr]
+	key := ep.Addr.Addr()
+
+	// Create or get the heap for this endpoint's addr
+	epHeap, found := et.endpoints[key]
 	if found {
-		entry.until = until
+		// Find the entry for this exact address; this loop is quick
+		// since we bound the number of items in the heap.
+		found := slices.ContainsFunc(epHeap, func(v *endpointTrackerEntry) bool {
+			return v.endpoint == ep
+		})
+		if !found {
+			heap.Push(&epHeap, &endpointTrackerEntry{endpoint: ep, until: until})
+		}
 	} else {
-		entry = endpointTrackerEntry{ep, until}
+		// Create a new heap with this exact address
+		heap.Push(&epHeap, &endpointTrackerEntry{endpoint: ep, until: until})
 	}
```

Comment:
not sure `found` + separate code paths really matters here. `ContainsFunc` on an empty slice will quickly report false. Could collapse this all a bit

Reply:
Yep, that seems to work 👍